File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
def get_next_nodes(nodes, node, tree):
    next_nodes = nodes.copy()
    next_nodes.remove(node)

    for child in tree[node]:
        next_nodes.add(child)

    return next_nodes


def dfs(nodes, tree, info, sheep, wolf):
    max_sheep = sheep

    for node in nodes:
        next_sheep = sheep
        next_wolf = wolf

        if info[node] == 0:
            next_sheep += 1
        else:
            next_wolf += 1

        if next_sheep <= next_wolf:
            continue

        next_nodes = get_next_nodes(nodes, node, tree)
        max_sheep = max(max_sheep, dfs(next_nodes, tree, info, next_sheep, next_wolf))

    return max_sheep


def solution(info, edges):
    tree = [[] for _ in range(len(info))]

    for u, v in edges:
        tree[u].append(v)

    nodes = {0}

    return dfs(nodes, tree, info, 0, 0)

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {
    public int solution(int[] info, int[][] edges) {
        
        int n = info.length;

        List<Integer>[] tree = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            tree[i] = new ArrayList<>();
        }

        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];

            tree[u].add(v);
        }

        Set<Integer> nodes = new HashSet<>();
        nodes.add(0);

        return dfs(nodes, tree, info, 0, 0);
    }
    
    public int dfs(Set<Integer> nodes, List<Integer>[] tree, int[] info, int sheep, int wolf) {

        int max = sheep;

        for (int node : nodes) {

            int nextSheep = sheep;
            int nextWolf = wolf;

            if (info[node] == 0) {
                nextSheep++;
            } else {
                nextWolf++;
            }

            if(nextSheep <= nextWolf) continue;

            Set<Integer> nextNodes = getNextNodes(nodes, node, tree);
            
            max = Math.max(max, dfs(nextNodes, tree, info, nextSheep, nextWolf));
        }

        return max;
    }

    public Set<Integer> getNextNodes(Set<Integer> nodes, int node, List<Integer>[] tree) {

        Set<Integer> nextNodes = new HashSet<>(nodes);
        nextNodes.remove(node); //부모 노드는 제거

        for (int child : tree[node]) {
            nextNodes.add(child);
        }

        return nextNodes;
    }
}